### PR TITLE
kubelet: Get Node object from etcd after Conflict error

### PR DESCRIPTION
**Pull request by linxiulei**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This RP refines the node status update logic to reduce load on `apiserver` and `etcd`.

Specifically two major issues:

1. With default settings, `kubelet` hits apiserver cache every 10s for getting Node object, which is unnecessary if we trust local Node object cache
2. In current retries, `kubelet` hits apiserver etcd even the last error is irrelevant to stale cache (e.g. 429s due to apiserver overload). 

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

File: pkg/kubelet/kubelet_node_status.go [hunk 1]
```diff
@@ -535,9 +534,11 @@ func (kl *Kubelet) syncNodeStatus() {
 // updateNodeStatus updates node status to master with retries if there is any
 // change or enough time passed from the last sync.
 func (kl *Kubelet) updateNodeStatus(ctx context.Context) error {
+	var lastErr error
 	klog.V(5).InfoS("Updating node status")
 	for i := 0; i < nodeStatusUpdateRetry; i++ {
-		if err := kl.tryUpdateNodeStatus(ctx, i); err != nil {
+		if err := kl.tryUpdateNodeStatus(ctx, lastErr); err != nil {
+			lastErr = err
```
Comment (HirazawaUi) [inline pkg/kubelet/kubelet_node_status.go:541]:
```suggestion
		if err := kl.tryUpdateNodeStatus(ctx, try); err != nil {
			if err != nil && apierrors.IsConflict(err) {
				try = true
			}
```
I think it's better to input bool type parameters instead of error...

Comment (linxiulei) [reply pkg/kubelet/kubelet_node_status.go]: Done, tks

File: staging/src/k8s.io/component-helpers/node/util/status.go [hunk 2]
```diff
@@ -38,7 +38,7 @@ func PatchNodeStatus(c v1core.CoreV1Interface, nodeName types.NodeName, oldNode
 
 	updatedNode, err := c.Nodes().Patch(context.TODO(), string(nodeName), types.StrategicMergePatchType, patchBytes, metav1.PatchOptions{}, "status")
 	if err != nil {
-		return nil, nil, fmt.Errorf("failed to patch status %q for node %q: %v", patchBytes, nodeName, err)
+		return nil, nil, fmt.Errorf("failed to patch status %q for node %q: %w", patchBytes, nodeName, err)
```
Comment (marquiz) [inline staging/src/k8s.io/component-helpers/node/util/status.go:41]: nit: looks unrelated. Maybe a separate PR (fixing other similar cases in the kubelet code)?

File: pkg/kubelet/kubelet_node_status.go [hunk 3]
```diff
@@ -551,20 +552,17 @@ func (kl *Kubelet) updateNodeStatus(ctx context.Context) error {
 
 // tryUpdateNodeStatus tries to update node status to master if there is any
 // change or enough time passed from the last sync.
-func (kl *Kubelet) tryUpdateNodeStatus(ctx context.Context, tryNumber int) error {
-	// In large clusters, GET and PUT operations on Node objects coming
-	// from here are the majority of load on apiserver and etcd.
-	// To reduce the load on etcd, we are serving GET operations from
-	// apiserver cache (the data might be slightly delayed but it doesn't
-	// seem to cause more conflict - the delays are pretty small).
-	// If it result in a conflict, all retries are served directly from etcd.
```
Comment (marquiz) [inline pkg/kubelet/kubelet_node_status.go:560]: This was an educative comment on the apiserver load and usage of cached object. Could we something similar for the new code, e.g. in the method "docstring" comment or somewhere. WDYT?

Comment (linxiulei) [reply pkg/kubelet/kubelet_node_status.go:560]: Yeah, I think the comment is still relevant after the change. Let me rephrase a bit.

File: staging/src/k8s.io/component-helpers/node/util/status.go (hunk 2 above)
Comment (linxiulei) [reply staging/src/k8s.io/component-helpers/node/util/status.go:41]: It's related. `apierrors.IsConflict()` needs the error to be wrapped (i.e. "%w"). I also filed https://github.com/kubernetes/kubernetes/issues/123234 as I think this is a more general issue for whole code base. The conclusion in that issue seems to suggest only change it when the intent is clearly to wrap the error, which is the case here.

**Comment by linxiulei** [comment]
Thanks @HirazawaUi for referencing https://github.com/kubernetes/kubernetes/pull/123977

/cc @wojtek-t looks like you have merged https://github.com/kubernetes/kubernetes/pull/123977 including part of this PR, which is getting Node object from local cache. I have rebased it and would like to continue with the remaining part, which is to only get Node object from etcd when there was a conflict error. (e.g. timeout errors should not impact etcd).

File: pkg/kubelet/kubelet_node_status.go (hunk 3 above)
Comment (yujuhong) [reply pkg/kubelet/kubelet_node_status.go:551]: @wojtek-t could you help confirm whether the approach -- only get from etcd directly when `apierrors.IsConflict()`-- is desired? Thanks!

Comment (wojtek-t) [reply pkg/kubelet/kubelet_node_status.go:551]: I wasn't aware of this PR...

It is and in fact already happened in: https://github.com/kubernetes/kubernetes/pull/123977, but it's desired on the first attempt. If we failed to update it (which BTW in practice is rare), we should rather get from apiserver in such case.
I think that doing that only for IsConflict() will not bring any non-negligible impact and introduces another complexity, so I wouldn't do that.

And if we exclude that, I think this PR is effectively no-op [on top of the my PR merged few months ago]

Comment (yujuhong) [reply pkg/kubelet/kubelet_node_status.go:551]: Got it. Agree that the main benefit already happened with the previous PR. Given the unclear impact of this PR, @linxiulei maybe we can close this PR.

Comment (linxiulei) [reply pkg/kubelet/kubelet_node_status.go:551]: @yujuhong @wojtek-t Let me clarify the value of this PR. From my observation, especially in cases when a cluster is being upscaled rapidly, updates are likely to be failed with 429s (i.e. apiserver couldn't serve that much requests), then in current design the kubelet would get the Node object from etcd, which compounds the situation.

So the value of the PR is to avoid such cases and make whole system is recoverable rather than kubelets keeping inflicting overhead to apiserver.

File: pkg/kubelet/kubelet_node_status.go [hunk 4]
```diff
@@ -542,24 +544,25 @@ func (kl *Kubelet) updateNodeStatus(ctx context.Context) error {
...
-		opts := metav1.GetOptions{}
-		originalNode, err = kl.heartbeatClient.CoreV1().Nodes().Get(ctx, string(kl.nodeName), opts)
-	}
+	// To reduce the load on both apiserver and etcd, we are getting Node
+	// object from local cache, which is synchronized via informer/watch (the
+	// data might be slightly delayed but it doesn't seem to cause more
+	// conflict - the delays are pretty small).
+	// If it result in a conflict, the following retry will be served directly
+	// from etcd.
+	originalNode, err := kl.GetNode()
```
Comment (aojea) [inline pkg/kubelet/kubelet_node_status.go:556]: if you don't use the cache why do you need to get the node from the informer if you are getting it directly later ?

Comment (linxiulei) [reply pkg/kubelet/kubelet_node_status.go:556]: Just for code simplicity, otherwise I need to declare an empty Node object like:

```
node := v1.Node{}
if useCache {
    node = kl.GetNode()
} else {
    node = kl.client.CoreV1().Nodes().Get()
}
```

I assumed getting the node from the informer is very cheap here.

Comment (aojea) [reply pkg/kubelet/kubelet_node_status.go:556]: code simplicity as in removing lines? this will be compiled so not much benefit on having less lines, it also seems harder to ready and to understand

File: pkg/kubelet/kubelet_node_status.go (hunk 3 above)
Comment (aojea) [reply pkg/kubelet/kubelet_node_status.go:551]: are those 429 not sending Retry-after headers?
https://github.com/kubernetes/kubernetes/blob/5bf1e95541d90e37f6c6637b5b45d8783e7907aa/staging/src/k8s.io/apiserver/pkg/server/filters/with_retry_after.go#L55-L65

Comment (linxiulei) [reply pkg/kubelet/kubelet_node_status.go:551]: I don't think so, because regardless of how the node is retrieved from, the update may be failed with 429s, right? This failure is not because of staleness but apiserver's being overwhelmed. Getting node objects from etcd doesn't really help but make the situation worse.

Comment (wojtek-t) [reply pkg/kubelet/kubelet_node_status.go:551]: 429s are automatically retries (IIRC 10 times by default) as Antonio noted. 
I would like to avoid over-optimizing as it always introduces additional complexity - let's see how much the already merged improvement helps and make a more data-driven decision after that.

Comment (linxiulei) [reply pkg/kubelet/kubelet_node_status.go:551]: Well, then I think this PR can wait until we see the need. Thanks all!

**Comment by yujuhong** [comment]
/hold

Based on the discussion in https://github.com/kubernetes/kubernetes/pull/123125#discussion_r1625701185, we'll revisit this optimization if there's a clear need.
